### DISCUSSION TITLE
关于B站/bilibili弹幕/在线播放的探讨

### DISCUSSION THREAD

**Discussion by sunjanesy** [hooke007/mpv_PlayKit#308] [mpv 通用]
一般来说只要搭配下面的[油猴脚本]和[弹幕脚本]就能满足要求,两者是相辅相成的关系
提一点:弹幕脚本如果没设置python环境变量,则需要按MD说明里改成exe模式才能正常显示

https://github.com/LuckyPuppy514/Play-With-MPV
https://github.com/itKelis/MPV-Play-BiliBili-Comments
![脚本](https://github.com/hooke007/MPV_lazy/assets/104602119/dd2a3c87-641c-422a-a41e-213005a20c80)

本来很煽情的画面,那句自动生成的中文字幕"阿姨啊"确实让我没绷住,噗~

-------------------------------------------------------------------------

我想探讨的是把cookies存于本地,调用yt-dlp.exe解析链接的模式,并不需要自己的账号实际点进链接的模式

优点:
可以解析大部分视频和直播,可选择视频源和音频源质量,也不用担心"把号养废",因为不会产生实际的浏览记录,随意观看
缺点:
不能解析番剧,没有弹幕看,不过使用楼下[suger520HT](https://github.com/suger520HT)提供的第三方编译版yt-dlp.exe可看弹幕,仅剩无法解析番剧和无法看直播弹幕

下载完把log后缀改回exe就可以正常使用:> [yt-dlp.log](https://github.com/hooke007/MPV_lazy/files/11454397/yt-dlp.log) 

----------------------------------------------------------------------------

简单介绍下用mpv_lazy本地存放cookies的方法

```
#本段代码需直接写进mpv.conf中,加#就的代表默认不启用

#启用 ytdl_hook 脚本
ytdl-format = bestvideo[height<=?2160][vcodec!=?av01]+bestaudio/best
                                         # 直接传递给yt-dlp的视频格式/质量 https://github.com/yt-dlp/yt-dlp#filtering-formats
                                         # 限制选择 最高2160(或选择未知高度) 视频编码不是av01(或选择未知编码) 的 最佳视频 + 最佳音频/最佳
# ytdl-raw-options = cookies-from-browser=Edge
ytdl-raw-options = cookies="D:\Program Files\MPV\portable_config\cookies.txt"
                                         # 将自定义选项传递给 yt-dlp，多个选项用 半角逗号(,)隔开，不要乱加空格 https://mpv.io/manual/master/#options-ytdl-raw-options
                                         # 两种方式分别为导入Edge的cookies与读取本地文本cookies.txt
                                         # 也可用 --ytdl-raw-options-append= 追加(仅单个选项)
#ytdl-raw-options-append = force-ipv6=   # 通过IPv6建立所有连接
## 代理，两者都应该启用
#ytdl-raw-options-append = proxy=http://127.0.0.1:7890
                                         # 让 yt-dlp 使用代理
#http-proxy = http://127.0.0.1:7890      # 让 mpv 使用 http 代理
```
●from-browser模式

`# ytdl-raw-options = cookies-from-browser=Edge`

cookies-from-browser 支持主流的所有浏览器,Edge/Chrome/firefox都可以
前提是所选浏览器需要登录B站,仓库能搜到该方法,不多讲,此法是从酷安基友那学的

●cookies.txt模式

`ytdl-raw-options = cookies="D:\Program Files\MPV\portable_config\cookies.txt"`

只能用完整路径,相对路径无效,cookies.txt的具体参数一共需要4个,此方案自己折腾了好久才会
```
# Netscape HTTP Cookie File
#其中XXX需要改成自己账号Cookies的具体值,第五列的0代表Cookies的有效期
.bilibili.com	TRUE	/	FALSE	0	SESSDATA	XXX
.bilibili.com	TRUE	/	FALSE	0	b_nut	XXX
.bilibili.com	TRUE	/	FALSE	0	buvid3	XXX
.bilibili.com	TRUE	/	FALSE	0	sid	XXX
```
如果你不会获取,可以参考我之前为某脚本录制这张图,登录后按F12
![cookies](https://github.com/hooke007/MPV_lazy/assets/104602119/fc166cd8-a37d-4352-a416-40db3f55b2a9)
设置好,测试可以看到已不是480P的低画质源
![mpv](https://github.com/hooke007/MPV_lazy/assets/104602119/46255cfb-520f-439b-86d8-a6c096cb5cbe)

----------------------------------------------------------------------------

非mpv.conf启动加载

●绑定预设快捷键模式

如果不想默认加载,可以在profiles.conf设置预设一个加载项,如下
```
[ytdl-profile]
 profile-desc = 开启ytdl-Cookies
 ytdl-format = bestvideo[height<=?2160][vcodec!=?av01]+bestaudio/best
 ytdl-raw-options = cookies="D:\Program Files\MPV\portable_config\cookies.txt"
```
然后在你的input列表里注册一个快捷键,例如mvp_lazy原版的input_uosc.conf里加入

`Ctrl+F1       show-text "已启用ytdl-cookie模式"; apply-profile ytdl-profile       #! 工具 > 开启ytdl解析`

只有在启用后才能用带cookies的方式去解析,不用重启播放器
![快捷键模式](https://github.com/hooke007/MPV_lazy/assets/104602119/ac653380-c7db-458d-8c0c-d406ec37ddc8)

●自动匹配模式

建一个自动匹配的profiles.conf的预设选项,不用按快捷键,自动匹配
```
[BiliBili_ytdl]
 profile-desc = Blibili-ytdl
 profile-cond = path:match("bilibili.com/")~=nil
 profile-restore = copy
 ytdl-raw-options = cookies="D:\Program Files\MPV\portable_config\cookies.txt"
 ytdl-format = bestvideo[height<=?2160][vcodec!=?av01]+bestaudio/best
```

●lua脚本自动识别模式

```
--检查URL是否包含指定的域名,并设置对应的ytdl-format和ytdl-raw-options[根据自身需求修改,放置\portable_config\scripts文件夹]
function setCookiesForDomain(url)
  if string.match(url, "bilibili.com") then
    mp.set_property("ytdl-raw-options", "cookies=\"D:\\Program Files\\MPV\\portable_config\\cookies.txt\",sub-langs=\"all,-danmaku\"")
    mp.set_property("ytdl-format", "bestvideo[height<=?2160][vcodec!=?av01]+bestaudio/best")
    return true
  elseif string.match(url, "youtube.com") then
    mp.set_property("ytdl-raw-options", "cookies-from-browser=Edge")
    mp.set_property("ytdl-format", "bestvideo[height<=?1080][vcodec=?h264]+bestaudio/best")
    return true
  end
  return false
end
-- 监听加载视频的事件
mp.add_hook("on_load", 50, function()
  local url = mp.get_property("stream-open-filename")
  setCookiesForDomain(url) 
end)
```
原本是没必要上lua脚本的,只因上文中的`profile-cond`不知道用什么属性才能匹配域名,各种失败,一气之下写了此脚本
可用path匹配还是后来大佬[hooke007](https://github.com/hooke007)告诉我的,这么多方案应该总有一款适合你,再次感谢文中提到的两位大佬的帮助.

**Comment by hooke007** [comment]
你这个 `profile-cond = false` 是哪里创造的语法

非条件profile移除掉两行无意义选项
`profile-cond`
`profile-restore`

**Reply by sunjanesy** [reply]
> 你这个 `profile-cond = false` 是哪里创造的语法
> 非条件profile移除掉两行无意义选项 `profile-cond` `profile-restore`

我是打算匹配bilibili.com这个域名的，没有找到profile-cond是否支持domain和url这类的属性，想着是判断语法先写成false应该没问题。谢谢指导，我先删掉，估计也完不成了

于是写了一个lua脚本

**Reply by sunjanesy** [reply]
最终还是写了一个简单的lua脚本😂

**Reply by hooke007** [reply]
你既然知道匹配哪个属性了，为了么不直接在profile的条件里改

**Reply by sunjanesy** [reply]
> 你既然知道匹配哪个属性了，为了么不直接在profile的条件里改

说实话，不是我不想，我拿stream-open-filename之类的都匹配过，试过了很多次，都失败了，官方文档里也没有类似案例，我也不知道能不能直接用mp.get_property("stream-open-filename")去匹配。最后想着试着写一个lua脚本得了

**Reply by Lonely233233** [reply]
[yt-dlp.log](https://github.com/hooke007/MPV_lazy/files/11454397/yt-dlp.log)
尝试下 一个别人重新编译的yt-dlp,改下扩展名就可以了，支持B站弹幕

**Reply by Lonely233233** [reply]
把扩展名重新改回EXE

**Reply by hooke007** [reply]
https://github.com/hooke007/MPV_lazy/discussions/42#discussioncomment-1680097

**Reply by sunjanesy** [reply]
> 把扩展名重新改回EXE

非常感谢，试了确实可以。
但是无法加上cookies了，要么有cookies，要么没字幕，请问有办法解决么

**Reply by sunjanesy** [reply]
> [#42 (comment)](https://github.com/hooke007/MPV_lazy/discussions/42#discussioncomment-1680097)

原来path就可以匹配，不过也算了解lua的基本写法了。

顺便说一句，这里真没人吐槽speed,lua布局问题么，既不居中，还能阻挡播放按键。我感觉要么再往下放点，在要么和声音一样竖着放，放在左边，鼠标移到左边才会显示
![MPV-00_04_59 920-N0002](https://github.com/hooke007/MPV_lazy/assets/104602119/6cea0218-8e6e-43eb-96a6-1b402f36899b)

**Reply by hooke007** [reply]
这和speed lua 有什么关系。。。
本来就没有绝对居中的设计，只是放在两侧控件间隔的中间而已。
想要类似绝对居中的设计，首先要静态图标，其次要左右间隔相等，想要实现可以自己重新设计每个按钮

> 还能阻挡播放按键

这是什么意思，播放按钮在最左侧，阻挡？？？

**Reply by sunjanesy** [reply]
> 这是什么意思，播放按钮在最左侧，阻挡？？？

我是说黑色倍速播放条，把蓝色按键给挡住了。

> 这和speed.lua 有什么关系。。。

我以为布局也是这个，我刚刚增加格子就是在这里，也在配置文件改了倍速，设置成0.05倍一格

**Reply by hooke007** [reply]
> 我是说黑色倍速播放条，把蓝色按键给挡住了。

我不是很理解你在做什么，osc_plus 和 uosc 本就是功能重合的不兼容脚本。这两也都是原始osc的替代品
脚本间的冲突都在此处有记录 https://github.com/hooke007/MPV_lazy/wiki/2_portable_config#%E5%A4%96%E7%BD%AE%E8%84%9A%E6%9C%AC%E8%B7%AF%E5%BE%84-

> 我刚刚增加格子就是在这里

我也不是很理解为什么要频繁修改脚本而不是改脚本选项。。。

**Reply by sunjanesy** [reply]
> osc_plus 和 uosc 本就是功能重合的不兼容脚本

原来不建议一起用，我一直看着这项目首页的两张图，以为不播放就是原版，播放时就会显示osc_plus，这样操作更方便。
除了这个倍速条不和谐，其他的我还真没感受到异样

**Reply by hooke007** [reply]
你还是没看懂，，不是不建议，是不应该。。。
不管是uosc 还是xxosc 这些都是osc，只能有一个。
这些替换脚本的存在是给人不同的选择，不是共存用的

**Reply by sunjanesy** [reply]
> 我也不是很理解为什么要频繁修改脚本而不是改脚本选项。。。

其实这真的没啥不好理解的，我改倍数确实是在uosc.conf里改的。后来打开脚本更改是因为我想尝试理解折腾。

正如我第一次找不到[default_directory=~/]被大佬批判一番，因为当时刚刚换到mpv-lazy标准包，里面只有脚本，没有script-opts这个配置文件夹，我就一直盯着脚本代各种找，后来下载这个文件夹才得以解决。

我知道项目本身是学习为主，文件不能覆盖，有很多部分是冲突，其他的我都不冲突，唯独这个我感觉不怎么影响，就没在意。

**Reply by hooke007** [reply]
说点题外话。
你可以看到obs这个版本和当前的lazy包完全是不同的风格（再往前的版本我甚至直接把中文离线文档塞进包里的）。为什么这么改的原始理由很简单，写的如obs那般详细也几乎没人看xxx.conf（也可能是沉默的大多数，但是我收到的不看的反馈实在太多），既然如此不如一切从简，只留链接指向文档出处（事实证明仍然是很多人连mpv.conf都不看的不然不会不知道文档在哪。。。）而且对我维护也很轻松。

说实话只靠官方的文档摸索全部的使用方式太难了（写的很详细但是选项分类并不是按照一般人的使用模式分的），因此创建了自己理解的文档 —— 我的文档是不是显得又杂又多，是的但是没办法，mpv的“乐高”属性太强，我尽量重新梳理结构了，谁有更好的点子尽管提。

**Reply by sunjanesy** [reply]
> [yt-dlp.log](https://github.com/hooke007/MPV_lazy/files/11454397/yt-dlp.log) 尝试下 一个别人重新编译的yt-dlp,改下扩展名就可以了，支持B站弹幕

经过反复测试这个只能在cookies-from-browser的情况下生效，本地cookies不兼容，不知道编译者有没有解决这个问题。

**Reply by sunjanesy** [reply]
> 说实话只靠官方的文档摸索全部的使用方式太难了（写的很详细但是选项分类并不是按照一般人的使用模式分的）

非常赞同，哪怕我花了好多个晚上就折腾mpv，至今也一知半解，文档多，配置文件也多，一般人都去打游戏追剧去了，才不会花费时间和精力学习这种内容分散又没有大用的东西，双击播放就行，配置文件是什么根本不会去在意。不是不反馈，是不知道从何反馈，而且如今上个GitHub都越来越难。

说实话osc_plus确实比原版好看，建议后续打包，可以旧版新版分开打包两个，根据需求选。一般人想换也无力自己解决。

**Reply by hooke007** [reply]
> 可以旧版新版分开打包两个

现在这几个包我打的都嫌累。。。
除非谁要能搞个osc切换脚本，我是很乐意集成进去

**Reply by Lonely233233** [reply]
> > [yt-dlp.log](https://github.com/hooke007/MPV_lazy/files/11454397/yt-dlp.log) 尝试下 一个别人重新编译的yt-dlp,改下扩展名就可以了，支持B站弹幕
> 
> 经过反复测试这个只能在cookies-from-browser的情况下生效，本地cookies不兼容，不知道编译者有没有解决这个问题。

这是别人重新编译的我也无能为力 但是他支持B站弹幕

**Reply by sunjanesy** [reply]
> > 可以旧版新版分开打包两个
> 
> 现在这几个包我打的都嫌累。。。 除非谁要能搞个osc切换脚本，我是很乐意集成进去


其实不用很复杂,比如默认用新版[portable_config]文件夹,旧版就用[portable_config旧版]文件夹，设置一个rename的bat,重命名文件夹就能一键切换,这样万一把哪边设置玩崩了，还有备用的。或是通过bat简单但移动配置文件就能实现

**Reply by sunjanesy** [reply]
> 谁要能搞个osc切换脚本，我是很乐意集成进去

要是需要的话，我就能做，比如在portable_config设置一个存放切换文件的文件夹，毕竟大多数文件不用换，batch只负责移进移出lua和conf文件

**Reply by hooke007** [reply]
并不是这个意思，以我的标准，要对两套ui做对应的体系设计。uosc一个脚本干了十个脚本的活，如果用osc_plus就需要把对应的缺失功能补上，这里边进一步涉及到对部分脚本的重构。但是我现在懒了不想做了，所以那些被取代的脚本只是放仓库，有人报bug就修，否则连新功能都不想试。假设双版本做出来，是不是还要完善文档对应新的用法。。。

**Reply by sunjanesy** [reply]
哈，那就没办法了，要是做出来，我倒是乐意帮忙测试或是完善部分的中文备注

**Reply by sunjanesy** [reply]
> 这是别人重新编译的我也无能为力 但是他支持B站弹幕

其实这基本上解决了需求，不知道有没有编译者发布地址，我去问问作者，很感谢

**Reply by Lonely233233** [reply]
![image](https://github.com/hooke007/MPV_lazy/assets/106970209/07aa02e3-a05e-4ad4-9c98-405bf79a9abe)
我没有登录也可以啊

**Reply by Lonely233233** [reply]
> > 这是别人重新编译的我也无能为力 但是他支持B站弹幕
> 
> 其实这基本上解决了需求，不知道有没有编译者发布地址，我去问问作者，很感谢

![image](https://github.com/hooke007/MPV_lazy/assets/106970209/41d1c61a-4822-490c-bab3-9774a9b40c8c)
我把ytdl-raw-options-append=cookies-from-browser=edge注释掉也是能加载弹幕的

**Reply by sunjanesy** [reply]
>  我把ytdl-raw-options-append=cookies-from-browser=edge注释掉也是能加载弹幕的

似乎你忘记了为什么要cookies,因为画质源,登录才有高清源,而不是480p,你看你两张图都是480p

**Reply by Lonely233233** [reply]
你不是说danmu吗

**Reply by sunjanesy** [reply]
>经过反复测试这个只能在cookies-from-browser的情况下生效，本地cookies不兼容，不知道编译者有没有解决这个问题
>> 你不是说danmu吗

我说的本地cookies.txt模式,加载弹幕不正常

发现是因为cookies.txt不支持ytdl-raw-options-append

我刚刚试了另一台电脑,似乎又可以直接用下面代码,又正常了
ytdl-raw-options = cookies="D:\Program Files\MPV\portable_config\cookies.txt"

**Comment by sunjanesy** [comment]
经测试可以正常匹配显示弹幕了,唯一的遗憾就是不能解析番剧,普通up的视频和直播都正常
```
[BiliBili_ytdl]
 profile-desc = Blibili-ytdl
 profile-cond = path:match("bilibili.com/")~=nil
 profile-restore = copy
 ytdl-raw-options = cookies="D:\Program Files\MPV\portable_config\cookies.txt"
 ytdl-format = bestvideo[height<=?2160][vcodec!=?av01]+bestaudio/best
```

**Reply by Lonely233233** [reply]
那是yt-dlp的问题

**Reply by sunjanesy** [reply]
> 那是yt-dlp的问题

明白，我之前用DS系某插件也是类似，不过是可以播放番剧的，只是没有高画质

**Reply by Lonely233233** [reply]
你去yt-dlp反馈 看看他们会不会修复

**Reply by sunjanesy** [reply]
> 你去yt-dlp反馈 看看他们会不会修复

这个估计不会修复，涉及版权保护啥的，真要看番剧上文章开头说的脚本就行，只是好像没有什么插件能让播放器支持实时弹幕除非独立开发的应用

**Reply by Lonely233233** [reply]
那就没办法了

**Comment by sunjanesy** [comment]
> 那就没办法了

据说新版已经修复了番剧播放问题，不知道第三方有没有跟进弹幕版

**Comment by Lonely233233** [comment]
![image](https://github.com/hooke007/MPV_lazy/assets/106970209/2e561f1b-5666-42a8-b076-81ff8d106538)
我现在能正常使用了

> > 那就没办法了
> 
> 据说新版已经修复了番剧播放问题，不知道第三方有没有跟进弹幕版

**Reply by WeiLaiF** [reply]
我用了上面那个yt还是不能放番咋整，普通up和直播都正常

**Reply by Lonely233233** [reply]
[yt-dlp.zip](https://github.com/hooke007/MPV_lazy/files/13427501/yt-dlp.zip)

**Reply by WeiLaiF** [reply]
这个链接我打开还是这个页面咋办，没下载弹窗。。

**Reply by hooke007** [reply]
因为这个链接就是指向这个帖子。可能他没传完就发言了，或者编辑错了。

**Comment by Lonely233233** [comment]


**Reply by hooke007** [reply]
你直接传zip不就行了，为什么要搞改名的操作（

**Reply by Lonely233233** [reply]
我不知道啊

**Reply by Lonely233233** [reply]
可以直接传ZIP吗

**Reply by Lonely233233** [reply]
早知道我早就传zip了

**Reply by hooke007** [reply]
你甚至可以在个人的仓库随意开个release的assets里发附件，没有大小和格式限制，在这个帖子里只需要留个地址指向。
这样就不必一直往楼层里加附件，在帖子里传过的附件是无法删除的，地址永远存在。你传assets里就不一样了，完全可控

**Reply by Lonely233233** [reply]
主要这个不是我弄的 我维护不了 这个也是别人给我的

**Reply by hooke007** [reply]
这跟是谁维护的没关系，我说的是gh传（二进制）文件的正确方式，如果你想把某个文件共享的话，最好就发自己仓库的release里

**Reply by Lonely233233** [reply]
谢谢大佬指教

**Reply by Lonely233233** [reply]
我有空多学习学习github的用法

**Reply by hooke007** [reply]
大概只需要三分钟吧。。https://docs.github.com/zh/repositories/releasing-projects-on-github/managing-releases-in-a-repository （这个中文翻译依托答辩，，英文只要不太差都只建议看英文版）

**Reply by WeiLaiF** [reply]
你那个yt-dlp有原帖嘛，能指个路吗😂

**Reply by Lonely233233** [reply]
没有都是别人传给我的

**Reply by WeiLaiF** [reply]
能再发一下那个yt-dlp嘛，找了一天没找到可以播番剧也可以加载弹幕的yt文件

**Reply by Lonely233233** [reply]
[yt-dlp.zip](https://github.com/hooke007/MPV_lazy/files/13427486/yt-dlp.zip)

**Reply by hooke007** [reply]
你这个文件没传完，你可以编辑看一下实际的链接

**Reply by WeiLaiF** [reply]
谢谢了，果然很强，又可以看番和日剧还能加载弹幕

**Reply by Lonely233233** [reply]
这东西都是别人给我 我也就借花献佛而已

**Reply by kaixc** [reply]
> [yt-dlp.zip](https://github.com/hooke007/MPV_lazy/files/13427486/yt-dlp.zip)

失效了，能不能看下作者有没有更新，现在解析弹幕出错了

**Reply by Lonely233233** [reply]
> > [yt-dlp.zip](https://github.com/hooke007/MPV_lazy/files/13427486/yt-dlp.zip)
> 
> 失效了，能不能看下作者有没有更新，现在解析弹幕出错了
[yt-dlp.zip](https://github.com/user-attachments/files/17178556/yt-dlp.zip)
 试试这个

**Reply by kaixc** [reply]
![image](https://github.com/user-attachments/assets/9be6e421-b72d-4e7d-b412-61b94f96e15d)
可以了，多谢

**Comment by MiKoto-Railgun** [comment]
> 只能用完整路径,相对路径无效,cookies.txt的具体参数一共需要4个,此方案自己折腾了好久才会
> 
> ```# Netscape HTTP Cookie File
> #其中XXX需要改成自己账号Cookies的具体值,第五列的0代表Cookies的有效期
> .bilibili.com	TRUE	/	FALSE	0	SESSDATA	XXX
> .bilibili.com	TRUE	/	FALSE	0	b_nut	XXX
> .bilibili.com	TRUE	/	FALSE	0	buvid3	XXX
> .bilibili.com	TRUE	/	FALSE	0	sid	XXX

现在好像得多加一个参数才能观看4K视频了

`.bilibili.com	TRUE	/	FALSE	0	bili_jct	XXX`
